File: android/app/src/main/java/de/maxifrz/lernwerk/pdf/DocxRenderer.kt
package de.maxifrz.lernwerk.pdf

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.graphics.pdf.PdfDocument
import android.text.Layout
import android.text.SpannableStringBuilder
import android.text.StaticLayout
import android.text.TextPaint
import android.text.style.StyleSpan
import de.maxifrz.lernwerk.present.DocxBlock
import de.maxifrz.lernwerk.present.DocxCell
import de.maxifrz.lernwerk.present.DocxDocument
import de.maxifrz.lernwerk.present.DocxRun
import java.io.ByteArrayOutputStream
import kotlin.math.max
import kotlin.math.min

/** Draws a parsed Word document onto A4 PDF pages: headings, paragraphs, lists, tables and pictures, paginated to
 * fit. Bold and italic are kept via Android's own text layout, so wrapping is automatic; other run formatting
 * (color, underline, font, exact spacing) is not read, and headers, footers and footnotes are left out. Mirrors
 * the iOS app. */
object DocxRenderer {
    private const val PAGE_WIDTH = 595
    private const val PAGE_HEIGHT = 842
    private const val MARGIN = 48f
    private val CONTENT_WIDTH = (PAGE_WIDTH - MARGIN * 2).toInt()

    fun pdfData(document: DocxDocument): ByteArray {
        val pdf = PdfDocument()
        val state = PageState(pdf)
        state.newPage()
        for (block in document.blocks) draw(block, state)
        state.finish()
        return ByteArrayOutputStream().also { pdf.writeTo(it); pdf.close() }.toByteArray()
    }

    /** The current PDF page and where drawing has reached on it; pages are started lazily as content needs them. */
    private class PageState(private val pdf: PdfDocument) {
        private var page: PdfDocument.Page? = null
        private var pageNumber = 0
        var canvas: Canvas? = null
            private set
        var y = MARGIN

        fun newPage() {
            page?.let(pdf::finishPage)
            pageNumber += 1
            val next = pdf.startPage(PdfDocument.PageInfo.Builder(PAGE_WIDTH, PAGE_HEIGHT, pageNumber).create())
            page = next
            canvas = next.canvas
            y = MARGIN
        }

        /** Starts a new page when `height` no longer fits, unless the page is still empty (an oversized block is
         * drawn as well as it can be rather than looping forever). */
        fun ensureSpace(height: Float) {
            if (y > MARGIN && y + height > PAGE_HEIGHT - MARGIN) newPage()
        }

        fun finish() {
            page?.let(pdf::finishPage)
        }
    }

    private fun draw(block: DocxBlock, state: PageState) {
        when (block) {
            is DocxBlock.Heading -> drawText(runsLayout(block.runs, style(block.level)), spacingBefore = if (block.level <= 1) 20f else 14f, state = state)
            is DocxBlock.Paragraph -> {
                if (block.runs.isEmpty()) {
                    state.ensureSpace(12f)
                    state.y += 12f
                    return
                }
                drawText(runsLayout(block.runs, bodyStyle), spacingBefore = 4f, state = state)
            }
            is DocxBlock.ListItem -> {
                val prefix = if (block.ordered) "${block.index}.  " else "•  "
                val inset = 14f + min(block.level, 4) * 16f
                drawText(runsLayout(block.runs, bodyStyle, prefix = prefix, maxWidth = CONTENT_WIDTH - inset.toInt()), spacingBefore = 3f, leftInset = inset, state = state)
            }
            is DocxBlock.Table -> drawTable(block.rows, block.columnWidths, state)
            is DocxBlock.Picture -> drawImage(block.data, block.aspect, state)
        }
    }

    private fun drawText(layout: StaticLayout, spacingBefore: Float, leftInset: Float = 0f, state: PageState) {
        state.ensureSpace(layout.height.toFloat())
        state.y += spacingBefore
        val canvas = state.canvas ?: return
        canvas.save()
        canvas.translate(MARGIN + leftInset, state.y)
        layout.draw(canvas)
        canvas.restore()
        state.y += layout.height + 3f
    }

    private fun drawImage(data: ByteArray, aspect: Double, state: PageState) {
        val bitmap = runCatching { BitmapFactory.decodeByteArray(data, 0, data.size) }.getOrNull() ?: return
        val ratio = if (aspect > 0) aspect.toFloat() else bitmap.width.toFloat() / max(bitmap.height, 1)
        var width = min(CONTENT_WIDTH.toFloat(), if (bitmap.width > 0) bitmap.width.toFloat() else CONTENT_WIDTH.toFloat())
        var height = width / max(ratio, 0.01f)
        val maxHeight = PAGE_HEIGHT - MARGIN * 2
        if (height > maxHeight) {
            height = maxHeight
            width = height * ratio
        }
        state.ensureSpace(height)
        state.y += 6f
        state.canvas?.drawBitmap(bitmap, null, RectF(MARGIN, state.y, MARGIN + width, state.y + height), Paint(Paint.FILTER_BITMAP_FLAG))
        state.y += height + 6f
    }

    private fun drawTable(rows: List<List<DocxCell>>, columnWidths: List<Double>, state: PageState) {
        val columnCount = rows.maxOfOrNull { it.size } ?: return
        if (columnCount == 0) return
        var widths = columnWidths.map { it.toFloat() }
        val total = widths.sum()
        widths = if (widths.size != columnCount || total <= 0) {
            List(columnCount) { CONTENT_WIDTH.toFloat() / columnCount }
        } else if (total != CONTENT_WIDTH.toFloat()) {
            val scale = CONTENT_WIDTH / total
            widths.map { it * scale }
        } else {
            widths
        }
        val padding = 5f
        val border = Paint().apply { color = Color.rgb(166, 166, 166); style = Paint.Style.STROKE; strokeWidth = 0.75f }
        state.y += 6f
        for (row in rows) {
            var rowHeight = 20f
            val cellLayouts = row.mapIndexed { index, cell ->
                val width = (widths.getOrNull(index) ?: widths.lastOrNull() ?: CONTENT_WIDTH.toFloat() / columnCount)
                val layout = cellLayout(cell, max(1, (width - padding * 2).toInt()))
                rowHeight = max(rowHeight, layout.height + padding * 2)
                layout to width
            }
            state.ensureSpace(rowHeight)
            var x = MARGIN
            val canvas = state.canvas
            for ((layout, width) in cellLayouts) {
                if (canvas != null) {
                    canvas.drawRect(x, state.y, x + width, state.y + rowHeight, border)
                    canvas.save()
                    canvas.translate(x + padding, state.y + padding)
                    layout.draw(canvas)
                    canvas.restore()
                }
                x += width
            }
            state.y += rowHeight
        }
        state.y += 8f
    }

    private fun cellLayout(paragraphs: DocxCell, width: Int): StaticLayout {
        val text = SpannableStringBuilder()
        paragraphs.forEachIndexed { index, runs ->
            if (index > 0) text.append('\n')
            appendRuns(text, runs)
        }
        if (text.isEmpty()) text.append(' ')
        return buildLayout(text, bodyStyle, width)
    }

    // Text layout

    private data class Style(val size: Float, val bold: Boolean)

    private val bodyStyle = Style(11f, bold = false)
    private fun style(level: Int) = Style(if (level <= 1) 22f else if (level == 2) 18f else 15f, bold = true)

    private fun runsLayout(runs: List<DocxRun>, style: Style, prefix: String = "", maxWidth: Int = CONTENT_WIDTH): StaticLayout {
        val text = SpannableStringBuilder()
        if (prefix.isNotEmpty()) {
            val start = text.length
            text.append(prefix)
            text.setSpan(StyleSpan(Typeface.BOLD), start, text.length, 0)
        }
        if (runs.isEmpty() && prefix.isEmpty()) text.append(' ')
        appendRuns(text, runs)
        return buildLayout(text, style, maxWidth)
    }

    private fun appendRuns(text: SpannableStringBuilder, runs: List<DocxRun>) {
        for (run in runs) {
            val start = text.length
            text.append(run.text)
            val face = when {
                run.bold && run.italic -> Typeface.BOLD_ITALIC
                run.bold -> Typeface.BOLD
                run.italic -> Typeface.ITALIC
                else -> null
            }
            if (face != null) text.setSpan(StyleSpan(face), start, text.length, 0)
        }
    }

    private fun buildLayout(text: CharSequence, style: Style, width: Int): StaticLayout {
        val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.BLACK
            textSize = style.size
            typeface = if (style.bold) Typeface.DEFAULT_BOLD else Typeface.DEFAULT
        }
        return StaticLayout.Builder.obtain(text, 0, text.length, paint, max(1, width))
            .setAlignment(Layout.Alignment.ALIGN_NORMAL)
            .setLineSpacing(3f, 1f)
            .build()
    }
}

File: Lernwerk/Services/Present/DocxRenderer.swift
import UIKit

/// Draws a parsed Word document onto A4 PDF pages: headings, paragraphs, lists, tables and pictures, paginated to
/// fit. Bold and italic are kept; other run formatting (color, underline, font, exact spacing) is not, and headers,
/// footers and footnotes are left out. Mirrors the Android app.
enum DocxRenderer {
    private static let pageSize = PaperRenderer.a4
    private static let margin: CGFloat = 48
    private static var contentWidth: CGFloat { pageSize.width - margin * 2 }

    static func pdfData(_ document: DocxDocument) -> Data {
        UIGraphicsPDFRenderer(bounds: CGRect(origin: .zero, size: pageSize)).pdfData { context in
            var y = margin
            context.beginPage()
            for block in document.blocks {
                draw(block, context: context, y: &y)
            }
        }
    }

    private static func draw(_ block: DocxBlock, context: UIGraphicsPDFRendererContext, y: inout CGFloat) {
        switch block {
        case let .heading(level, runs):
            drawText(attributed(runs, style: .heading(level)), spacingBefore: level <= 1 ? 20 : 14, context: context, y: &y)
        case let .paragraph(runs):
            guard !runs.isEmpty else {
                ensureSpace(12, context: context, y: &y)
                y += 12
                return
            }
            drawText(attributed(runs, style: .body), spacingBefore: 4, context: context, y: &y)
        case let .listItem(level, ordered, index, runs):
            let prefix = ordered ? "\(index).\t" : "•\t"
            drawText(attributed(runs, style: .body, prefix: prefix), spacingBefore: 3, leftInset: 14 + CGFloat(min(level, 4)) * 16, context: context, y: &y)
        case let .table(rows, columnWidths):
            drawTable(rows, columnWidths: columnWidths, context: context, y: &y)
        case let .image(data, aspect):
            drawImage(data, aspect: aspect, context: context, y: &y)
        }
    }

    /// Starts a new page when `height` no longer fits, unless the page is still empty (an oversized block is
    /// drawn as well as it can be rather than looping forever).
    private static func ensureSpace(_ height: CGFloat, context: UIGraphicsPDFRendererContext, y: inout CGFloat) {
        if y > margin, y + height > pageSize.height - margin {
            context.beginPage()
            y = margin
        }
    }

    private static func drawText(_ text: NSAttributedString, spacingBefore: CGFloat, leftInset: CGFloat = 0, context: UIGraphicsPDFRendererContext, y: inout CGFloat) {
        let width = contentWidth - leftInset
        guard width > 20 else { return }
        let bounds = text.boundingRect(with: CGSize(width: width, height: .greatestFiniteMagnitude), options: [.usesLineFragmentOrigin, .usesFontLeading], context: nil)
        let height = ceil(bounds.height)
        ensureSpace(height, context: context, y: &y)
        y += spacingBefore
        text.draw(with: CGRect(x: margin + leftInset, y: y, width: width, height: max(height, pageSize.height)), options: [.usesLineFragmentOrigin, .usesFontLeading], context: nil)
        y += height + 3
    }

    private static func drawImage(_ data: Data, aspect: Double, context: UIGraphicsPDFRendererContext, y: inout CGFloat) {
        guard let image = UIImage(data: data) else { return }
        let aspect = aspect > 0 ? CGFloat(aspect) : image.size.width / max(image.size.height, 1)
        var width = min(contentWidth, image.size.width > 0 ? image.size.width : contentWidth)
        var height = width / max(aspect, 0.01)
        let maxHeight = pageSize.height - margin * 2
        if height > maxHeight {
            height = maxHeight
            width = height * aspect
        }
        ensureSpace(height, context: context, y: &y)
        y += 6
        image.draw(in: CGRect(x: margin, y: y, width: width, height: height))
        y += height + 6
    }

    private static func drawTable(_ rows: [[DocxCell]], columnWidths: [Double], context: UIGraphicsPDFRendererContext, y: inout CGFloat) {
        let columnCount = rows.map(\.count).max() ?? 0
        guard columnCount > 0 else { return }
        var widths = columnWidths.map { CGFloat($0) }
        let total = widths.reduce(0, +)
        if widths.count != columnCount || total <= 0 {
            widths = Array(repeating: contentWidth / CGFloat(columnCount), count: columnCount)
        } else if total != contentWidth {
            let scale = contentWidth / total
            widths = widths.map { $0 * scale }
        }
        let padding: CGFloat = 5
        y += 6
        for row in rows {
            var cellTexts: [(text: NSAttributedString, width: CGFloat)] = []
            var rowHeight: CGFloat = 20
            for (index, cell) in row.enumerated() {
                let width = widths.indices.contains(index) ? widths[index] : (widths.last ?? contentWidth / CGFloat(columnCount))
                let text = cellText(cell)
                let bounds = text.boundingRect(with: CGSize(width: width - padding * 2, height: .greatestFiniteMagnitude), options: [.usesLineFragmentOrigin], context: nil)
                rowHeight = max(rowHeight, ceil(bounds.height) + padding * 2)
                cellTexts.append((text, width))
            }
            ensureSpace(rowHeight, context: context, y: &y)
            var x = margin
            for (text, width) in cellTexts {
                let rect = CGRect(x: x, y: y, width: width, height: rowHeight)
                UIColor(white: 0.65, alpha: 1).setStroke()
                UIBezierPath(rect: rect).stroke()
                text.draw(with: rect.insetBy(dx: padding, dy: padding), options: [.usesLineFragmentOrigin], context: nil)
                x += width
            }
            y += rowHeight
        }
        y += 8
    }

    private static func cellText(_ paragraphs: DocxCell) -> NSAttributedString {
        let result = NSMutableAttributedString()
        for (index, runs) in paragraphs.enumerated() {
            if index > 0 { result.append(NSAttributedString(string: "\n")) }
            result.append(attributed(runs, style: .body))
        }
        if result.length == 0 { result.append(placeholder) }
        return result
    }

    // MARK: - Text attributes

    private enum Style {
        case heading(Int)
        case body
    }

    private static let placeholder = NSAttributedString(string: " ", attributes: [.font: font(bold: false, italic: false, style: .body)])

    private static func attributed(_ runs: [DocxRun], style: Style, prefix: String = "") -> NSAttributedString {
        let result = NSMutableAttributedString()
        if !prefix.isEmpty {
            result.append(NSAttributedString(string: prefix, attributes: [.font: font(bold: true, italic: false, style: style)]))
        }
        if runs.isEmpty, prefix.isEmpty {
            result.append(placeholder)
        }
        for run in runs {
            result.append(NSAttributedString(
                string: run.text,
                attributes: [.font: font(bold: run.bold, italic: run.italic, style: style), .foregroundColor: UIColor.black]
            ))
        }
        return result
    }

    private static func font(bold: Bool, italic: Bool, style: Style) -> UIFont {
        let size: CGFloat
        let base: UIFont
        switch style {
        case let .heading(level):
            size = level <= 1 ? 22 : (level == 2 ? 18 : 15)
            base = UIFont(name: QuillFont.Weight.semibold.postScriptName, size: size) ?? .boldSystemFont(ofSize: size)
        case .body:
            size = 11
            let weight: QuillFont.Weight = bold ? .semibold : .regular
            base = UIFont(name: weight.postScriptName, size: size) ?? .systemFont(ofSize: size)
        }
        guard italic else { return base }
        let traits = base.fontDescriptor.symbolicTraits.union(.traitItalic)
        guard let descriptor = base.fontDescriptor.withSymbolicTraits(traits) else { return base }
        return UIFont(descriptor: descriptor, size: size)
    }
}
